bots: Add a setting to customize the Welcome Bot message.

saubhagya-patel opened the pull request:
This PR adds a setting that allows the Administrator to customize the Welcome Bot message upon invitation.

Fixes: #27663.
[CZO thread](https://chat.zulip.org/#narrow/stream/101-design/topic/Welcome.20Bot.20customization/near/1528559)
  
**Previous PR reference:**
- This PR continues the unfinished work done in PR [#29452](https://github.com/zulip/zulip/pull/29452) by @akarsh-jain-790.
- The developments made are as follows:
  - Updated the variable and file names to be consistent as suggested [here](https://github.com/zulip/zulip/pull/29452#discussion_r1613240043).
  - Correct the failing API tests.

<!-- Describe your pull request here.-->


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

  <details>
   <summary>Added option to customize the Welcome Bot message in the Invite modal.</summary>

| Configuration Status of custom Welcome message| Checkbox Selected | Checkbox Not Selected |
|----------------------|------------------------|--------------------|
| **Not Configured**   | <img width="757" height="1003" alt="image" src="https://github.com/user-attachments/assets/ce9a19e1-7b9e-434a-ad67-b944692a5304" /> | <img width="759" height="870" alt="image" src="https://github.com/user-attachments/assets/8e954454-fb6d-401c-8c60-e8e76fb2c4e6" /> |
| **Configured**       | ![Not Selected, Configured](https://github.com/user-attachments/assets/7d51709e-f5ec-4a7d-bd6a-a219716ef1b4) | ![Selected, Configured](https://github.com/user-attachments/assets/ad9807a4-054d-414a-ad87-b4644cf568b4) |

 **Non-Admin View**: 
 <img src="https://github.com/user-attachments/assets/696bcc1e-03fc-45c2-9bd0-fb93ecdf1dde" width="400" alt="Non-Admin View"/>

  </details>




  <details>
  <summary> Added setting in Organization settings. </summary>

  **Setting disabled:**
  ![image](https://github.com/user-attachments/assets/755c28a8-c69a-41d3-92a0-0e23bfffe95b)

  **Setting enabled:**
  ![image](https://github.com/user-attachments/assets/0d6c331a-cd36-448f-930b-d7e5c359be05)

  **Testing Welcome Bot custom message:**
  ![added view message button](https://github.com/user-attachments/assets/1f1db08e-f27b-46d8-87af-de8ac13e7c22)

  </details>
  
  <details>

   <summary>Welcome Bot message sent to users upon invitation.</summary>  
   
  **Welcome Bot custom message when included in an invitation.**
  
   ![image](https://github.com/user-attachments/assets/74b2008f-43db-4234-ab97-7c14bc1c07ab)
  
  ![image](https://github.com/user-attachments/assets/ec56fc7b-523f-4d74-a3e6-c0e826d756e3)
  
  **Welcome Bot custom message when not included in an invitation.**
   
   ![image](https://github.com/user-attachments/assets/65caebb8-d196-4fd5-917e-7dafb518c280)
   
   ![image](https://github.com/user-attachments/assets/56555cea-c94f-4765-9662-7bd98ad9e869)

  </details>
  
  
  <details>
  <summary> Generated API documentation for <code>/realm/test_welcome_bot_custom_message</code> </summary>
   
   
  ![Screenshot 2025-01-01 172401](https://github.com/user-attachments/assets/f0e829b8-4056-467f-9506-3201667720e5)
  
  
  ![Screenshot 2025-01-01 172232](https://github.com/user-attachments/assets/6b057d0d-7d06-47e0-aa12-990887916ed7)

   
  </details>

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

saubhagya-patel commented:
The change in the [`organization_settings_admin.hbs`](https://github.com/zulip/zulip/pull/32862/files#diff-04d42001bd1f60b9f4ac7f8ff7c70cf3f03f7d0a41d6cd698e44499d954112f8) - specifically removing the class `inline-block` from the `organization-settings-parent` was done to make sure that the width is applied properly to the `<textarea>` element in the Welcome Bot custom message setting.

<details>
<summary>Screenshot of the <code><textarea></code> element if the above change wasn't made.</summary>

![img](https://github.com/user-attachments/assets/890d5859-f961-41ce-ab5a-2bcbd25be02b)

</details>

<details>
<summary>Screenshot of the <code><textarea></code> element with the change made.</summary>

![img](https://github.com/user-attachments/assets/754b5f94-d245-41a9-a5e2-13cc6cb46c88)

</details>

Aditya8840 commented:
@saubhagya-patel  Please have a look: https://zulip.readthedocs.io/en/latest/contributing/continuing-unfinished-work.html#review-existing-work-and-feedback

alya commented:
Thanks! Testing the PR:

- The error and confirmation banners disappear way too quickly. Did you match some delay used elsewhere? Perhaps we can just leave the banner until the next time you click the test message button, or until you navigate away, whichever happens first.
- Markdown formatting in the message messes up the quoting. I expect you should use ` ```quote`, not `>`.

![Screenshot 2025-01-02 at 15 46 06](https://github.com/user-attachments/assets/c7fcec68-fdc5-41c2-9012-21f87dc5b312)

alya commented:
I think we should put the admin note into a separate message, rather than tacking it on to the end of the first welcome bot message.

saubhagya-patel commented on zerver/lib/onboarding.py:
```diff
@@ -99,14 +119,29 @@ def send_initial_direct_message(user: UserProfile) -> int:
+            internal_prep_private_message(
+                get_system_bot(settings.WELCOME_BOT, user.realm_id),
+                user,
+                welcome_bot_custom_message_string,
```
Here, I didn't use the `remove_single_newlines()` function because the administrator may want to keep the formatting on separate lines and it created an error while displaying the quoted content.

<details>
<summary>Screenshot of the welcome bot custom message with the <code>remove_single_newlines()</code> function.</summary>

![image](https://github.com/user-attachments/assets/e9ded99c-d016-4d45-865e-b8c02dd4dcc1)

</details>

<details>
<summary>Screenshot of the welcome bot custom message without the <code>remove_single_newlines()</code> function.</summary>

![image](https://github.com/user-attachments/assets/409cb6e3-c8bd-4384-8e58-99d32d91dfbd)

</details>

timabbott replied:
Yeah, this seems correct, `remove_single_newlines` is for a very specific composition UI.

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes are made:

- In the Welcome Bot custom message now ` ```quote{}``` ` is used instead of  `>` to preserve markdown formatting following [this](https://github.com/zulip/zulip/pull/32862#issuecomment-2568530875) comment.
  <details>
  <summary>Screenshot of the Welcome Bot custom message with markdown formatting.</summary>
  
   ![image](https://github.com/user-attachments/assets/a80062fb-6b4a-4819-971a-10b0212f16fa)

  </details>


- The Welcome Bot custom message is now put into a separate message following [this](https://github.com/zulip/zulip/pull/32862#issuecomment-2568532021) comment.
  <details>
  <summary>Screenshot of the Welcome Bot custom message put into a separate message.</summary>
  
   ![image](https://github.com/user-attachments/assets/955277c6-b763-40a8-a1b4-ece3a0dd3d95)

  </details>
  
- The error and confirmation banners are now not hidden automatically. They are hidden, when we toggle the Setting option "Send a custom Welcome Bot message to new users" or we exit the Settings overlay following [this](https://github.com/zulip/zulip/pull/32862#issuecomment-2568530875) comment.

  `Alya Abbott` [said](https://github.com/zulip/zulip/pull/32862#issuecomment-2568530875):
  > The error and confirmation banners disappear way too quickly. Did you match some delay used elsewhere?
  
  Apologies for the inconvenience. The issue occurred on my end when I mistakenly used the same value as it appears when saving changes in the Manage User or Manage Bot tabs. However, we do implement a delay of `4000ms` in certain places, such as when changing the email in the development environment (`settings_account.ts`) or when an export is initiated (`settings_exports.ts`).
  
@alya, this PR is ready for review. I hope the updated changes look good to you. Please let me know if any further adjustments are needed. Thank you.

alya commented:
Thanks! @sahil839 Could you please do the next round of review on this one? I haven't re-tested.

Not a 10.0 priority.

saubhagya-patel commented:
I have updated this PR following the discussion on the [CZO thread](https://chat.zulip.org/#narrow/channel/101-design/topic/Welcome.20Bot.20customization/near/2051273). The following changes are made:

- The Welcome Bot custom message can now be customized by the Administrator per invitation. Please have a look at the attached screenshots.

- The organization's default custom message (if set by the Administrator) is prepopulated in the invite modal. The Administrators can send a test message or save the default message from the invite modal.

- Visual changes: 

  <details>
   <summary>Added option to customize the Welcome Bot message in the Invite modal.</summary>

   | Description                                      | Dark Mode                                                                                     | Light Mode |
   |--------------------------------------------------|------------------------------------------------------------------------------------------------|------------|
   | **Checkbox not selected**                        | ![image](https://github.com/user-attachments/assets/949c7703-9395-44a8-b35b-23e67371113f)       | ![image](https://github.com/user-attachments/assets/a47337be-de3c-47c9-bda9-a760776e160d)            |
   | **Checkbox selected**                            | ![image](https://github.com/user-attachments/assets/2efcd331-b51d-4733-8094-584fb871ea9e)       | ![image](https://github.com/user-attachments/assets/4951157a-fb49-4e13-b614-9fd53195c949)            |
   | **Setting is not available for Non-Admin Users** | ![image](https://github.com/user-attachments/assets/18145934-609b-48f7-963a-a4324cca4a51)       | ![image](https://github.com/user-attachments/assets/8e93c593-68ba-49d5-812d-30b062de911e)            |

  </details>


  <details>
   <summary>Interaction with the option.</summary>
   
   **Saving a non-empty default Welcome Bot custom message.**

  ![default save (2)](https://github.com/user-attachments/assets/1499cc7e-81de-4169-92e0-f4037c90673b)
  
  **Saving an empty default Welcome Bot custom message fails.**
  
   ![default save empty](https://github.com/user-attachments/assets/4baa2f67-0a3a-4303-9141-c2e00113d5f9)

   **Testing a non-empty default Welcome Bot custom message.**
  
  ![test message](https://github.com/user-attachments/assets/de7dd1d2-23e1-45d7-b5a1-58702dd97702)

  **Testing an empty Welcome Bot custom message fails.**

   ![test message empty](https://github.com/user-attachments/assets/fad7cc85-f7f8-48d8-8ce7-b1c80900fd0c)
   
  </details>

  <details>

   <summary>Welcome Bot message sent to users upon invitation.</summary>  
   
   **Welcome Bot custom message when included in an invitation.**
  
  ![image](https://github.com/user-attachments/assets/508a6834-c218-4442-885f-ff617f1e2345)
  
  ![image](https://github.com/user-attachments/assets/790facb7-a725-4623-bebf-61b3634a6c73)
  
  **Welcome Bot custom message when not included in an invitation.**
   
   ![image](https://github.com/user-attachments/assets/1978de14-543b-4715-ade6-a49a0ca4a663)
   
   ![image](https://github.com/user-attachments/assets/33b1994a-4ba0-4a3d-a89b-7d2276271e35)

  </details>



@alya, this PR is ready for review. I hope the updated changes look good to you. Please let me know if any further changes are required. Tha.k you.

alya commented:
Why is "User groups they should join" missing from the screenshots above?

alya commented:
Please bump this after the release; I won't have time to review/test before then.

saubhagya-patel commented:
> Why is "User groups they should join" missing from the screenshots above?

This happened because when the database was populated, only one group was created, and the Administrator didn't have permission to add users to it. I have updated the screenshots. Thanks.

sahil839 commented:
@alya should I review this now or should I wait for you to test it once?

saubhagya-patel commented:
Upon reviewing this PR, I found out that I had forgotten to update the Zulip version from 10.0 to 11.0 in the `zulip.yaml` file. Hence, the above push was made. Thank you.

alya commented:
Please update the PR to use the strings provided in the issue.

alya commented:
Other notes from manual testing:

1. When the "Save changes" button is disabled, it needs a tooltip:
> Welcome Bot message text is required.

2. The message box should be much smaller both in settings and the invite modal; 3 lines by default should work.
3. For users who can't edit the test message, let's have the box fit the content if possible. That would be a nice improvement to make to the organization description box as well.
4. For the text message, instead of a confirmation banner, let's have a "View message" button appear next to the "Send me a test message" button. It should be a low-emphasis success button using new button styles, and take you to the test message (/near/ link if we can get that; DMs with Welcome Bot otherwise).

alya commented:
I'm removing the "maintainer review" label for now. Please re-add it and ping @sahil839 when ready.

saubhagya-patel commented:
Following the review received, I have updated the PR. The following changes were made:

- Following [this](https://github.com/zulip/zulip/pull/32862#issuecomment-2773232756) comment, I have updated the PR to use the strings as per the issue.

- Following [this](https://github.com/zulip/zulip/pull/32862#issuecomment-2773273717) comment, the following changes were made:
  - Added tooltip to "Save changes" button when disabled.
  - The message box is now 3 lines by default, in the settings as well as the invite modal.
  - Added "View message" button with the /near/ link.
  - I am trying to find a good solution for this, (will ping `Alya Abbott` once I can):
  
    > For users who can't edit the test message, let's have the box fit the content if possible. That would be a nice improvement to make to the organization description box as well.
  
- Additional changes made:
  - Added the same tooltip to the "Send me a test message" button when disabled.
  - If an error occurs, previously it was shown through the banner, now I changed it to show "Error sending message", just like the "View message" button.
  - Please refer to the screenshots for the visual changes made.

<details>
<summary>Screenshots</summary>

- Updated strings in the organization settings and the invite modal.
![image](https://github.com/user-attachments/assets/0d6c331a-cd36-448f-930b-d7e5c359be05)
![image](https://github.com/user-attachments/assets/31db6425-b9e2-46ed-bfaa-11f3f1faee7f)

- Added tooltip to "Save changes" button when disabled.
![image](https://github.com/user-attachments/assets/170a69ba-14bd-4abd-ac64-8cf709bc7766)

- Added "View message" button.
![image](https://github.com/user-attachments/assets/90f87d55-071e-440c-972a-956dcf042d23)
![added view message button](https://github.com/user-attachments/assets/1f1db08e-f27b-46d8-87af-de8ac13e7c22)

- Added the same tooltip to the "Send me a test message" button when disabled.
![image](https://github.com/user-attachments/assets/99446b5b-e654-472f-99a1-17007fda83a5)

- Updated display of error, if any.
![image](https://github.com/user-attachments/assets/e780b2cc-a7d9-4321-9512-39eb577efe20)
![updated error display](https://github.com/user-attachments/assets/bf146188-be18-4256-a28f-9e7db9277d79)

</details>




@alya, I have made the required changes.
Do you want to do a quick review of the UI changes, or should I ping `Sahil Batra` for the "maintainer review"?
Thank you.

alya commented:
I think @sahil839 can do the next round of review, thanks!

sahil839 commented on zerver/tests/test_invite.py:
```diff
@@ -2829,6 +2880,68 @@ def check_user_able_to_register(self, email: str, invite_link: str) -> None:
+                "welcome_bot_custom_message": welcome_bot_custom_message,
+            },
+        )
+        self.assert_json_error(result, "Insufficient permission")
```
I think there is no need to check this case, as we might be already testing the case where user does not have permission to create multiuse invites.

saubhagya-patel commented on zerver/lib/onboarding.py:
```diff
@@ -84,6 +90,18 @@ def send_initial_direct_message(user: UserProfile) -> int:
+            welcome_bot_custom_message_content = welcome_bot_custom_message
+            can_add_welcome_bot_custom_message = True
+        if can_add_welcome_bot_custom_message:
+            welcome_bot_custom_message_string = _("""
```
Sure, that sounds reasonable.

saubhagya-patel commented on zerver/lib/onboarding.py:
```diff
@@ -106,14 +124,29 @@ def send_initial_direct_message(user: UserProfile) -> int:
+    message_id = do_send_messages(welcome_bot_message)[0].message_id
+
     assert message_id is not None
     return message_id
```
Yes.
I asked on [CZO](https://chat.zulip.org) here: [#design > Welcome Bot customization @ 💬](https://chat.zulip.org/#narrow/channel/101-design/topic/Welcome.20Bot.20customization/near/2160259).
Based on the thread, we do need to star both (initial welcome bot message and custom Welcome Bot message), this is implemented in the following push.

saubhagya-patel commented on web/src/settings_components.ts:
```diff
@@ -1472,6 +1483,21 @@ function enable_or_disable_save_button($subsection_elem: JQuery): void {
+        disable_save_button = should_disable_save_button_for_welcome_bot_custom_message_setting();
+        if (disable_save_button) {
+            // avoid duplication of tippy
+            if (!tippy_instance) {
```
This code path was already updated in [0914732 settings: Improve enable_or_disable_save_button method code.](https://github.com/zulip/zulip/commit/09147323879a61055861ecf7952545f31fe5bc71). Hence, I don't think we need to update this now.

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes were made:

- Follwing [this](https://github.com/zulip/zulip/pull/32862#discussion_r2010036890) comment, to simplify `send_initial_direct_message()` function in `zerver/lib/onboarding.py`, it is updated.

- Following [this](https://github.com/zulip/zulip/pull/32862#discussion_r2010038903) comment, to use the message ID for starring the Welcome Bot custom message, I asked on [CZO](https://chat.zulip.org/) here: [#design > Welcome Bot customization @ 💬](https://chat.zulip.org/#narrow/channel/101-design/topic/Welcome.20Bot.20customization/near/2160259). Based on the thread, we do need to star both (initial welcome bot message and custom Welcome Bot message), hence now both the messages are starred.

- Following [this](https://github.com/zulip/zulip/pull/32862#discussion_r2044145423) comment, to deduplicate code to handle tippy instances, it was updated in [0914732 settings: Improve enable_or_disable_save_button method code](https://github.com/zulip/zulip/commit/09147323879a61055861ecf7952545f31fe5bc71). Hence, I don't think we need to update this now.

- Following [this](https://github.com/zulip/zulip/pull/32862#discussion_r2044147919) comment, `height` CSS is now added in `#id_realm_default_welcome_bot_custom_message`.

- Following [this](https://github.com/zulip/zulip/pull/32862#discussion_r2044154026) comment, the test for testing Welcome Bot custom message from a normal account, the test is removed as we already test the condition when the user doesn't have permission to create multiuse invites.

- <details>
  <summary> Visual changes (now the Welcome Bot custom message is also starred).  </summary>
  
  **Welcome Bot custom message when included in an invitation.**
  
  ![image](https://github.com/user-attachments/assets/74b2008f-43db-4234-ab97-7c14bc1c07ab)
  
  ![image](https://github.com/user-attachments/assets/ec56fc7b-523f-4d74-a3e6-c0e826d756e3)
  
  **Welcome Bot custom message when not included in an invitation.**
   
   ![image](https://github.com/user-attachments/assets/65caebb8-d196-4fd5-917e-7dafb518c280)
   
   ![image](https://github.com/user-attachments/assets/56555cea-c94f-4765-9662-7bd98ad9e869)

  </details>


@sahil839, I have made the required changes. I hope the updated changes look good to you. Please let me know if any further adjustments are required.
Thank you.

sahil839 commented on web/templates/invite_user_modal.hbs:
```diff
@@ -97,4 +97,19 @@
             </div>
         </div>
     {{/if}}
+    {{#if is_admin}}
```
Should we show this if the organization setting is not checked? I am not sure if this has been discussed before.

sahil839 commented on zerver/views/invite.py:
```diff
@@ -152,6 +165,9 @@ def invite_users_backend(
     ]
     check_role_based_permissions(invite_as, user_profile, require_admin=require_admin)
 
+    if welcome_bot_custom_message is not None and not user_profile.is_realm_admin:
```
How does organization setting control sending custom message during invites? Should we allow it if organization setting is disabled? Although admins are only allowed to include custom messages and they can modify the organization setting but still.

sahil839 reviewed (commented):
Posted a couple of comments. I am not sure how the organization setting is used here, so posted questions about that.

saubhagya-patel commented on zerver/views/invite.py:
```diff
@@ -273,5 +296,6 @@ def generate_multiuse_invite_backend(
         include_realm_default_subscriptions,
         streams,
         user_groups,
+        welcome_bot_custom_message,
```
No, only admins can configure the Welcome Bot custom message for email invites or multi-use invites.

saubhagya-patel commented:
Based on the thread [#design > Welcome Bot customization](https://chat.zulip.org/#narrow/channel/101-design/topic/Welcome.20Bot.20customization/with/2051273), we have the following setting system:

- There are two realm settings: `send_invite_welcome_bot_custom_message` and `default_welcome_bot_custom_message`. These are used as:
  - `send_invite_welcome_bot_custom_message`: a checkbox to visually hide the textarea (containing value of default Welcome Bot custom message) in the settings panel, and whether the Welcome Bot custom message checkbox is checked in the invite modal
  - `default_welcome_bot_custom_message`: the default value of Welcome Bot custom message
  
  Both settings can only be modified by organization admins.

- The checkbox in the invite modal is shown only for admins, as the Welcome Bot custom message is only allowed to be sent by admins. We don't show it to a Member or a Moderator.

@sahil839, this PR is already updated for the stated setting system, so I would request for another round of review.
Thanks.

alya commented:
To be clear, if `default_welcome_bot_custom_message` is configured, it should always be sent when a non-admin is creating the invitation.

saubhagya-patel commented on zerver/lib/onboarding.py:
```diff
@@ -110,16 +123,37 @@ def send_initial_direct_message(user: UserProfile) -> int:
+        welcome_bot_custom_message_id = sent_messages[1].message_id
+        return message_id, welcome_bot_custom_message_id
+    else:
+        return message_id, None
```
Sure, refactored the code and added necessary comments.

saubhagya-patel commented on web/src/settings_org.ts:
```diff
@@ -1346,6 +1433,8 @@ export function build_page(): void {
     set_create_web_public_stream_dropdown_visibility();
     disable_create_user_groups_if_on_limited_plan();
+    update_test_welcome_bot_custom_message_button_status();
+    set_welcome_message_custom_text_visibility();
```
Ohh, right, not needed here.

saubhagya-patel commented on zerver/views/invite.py:
```diff
@@ -137,6 +144,12 @@ def invite_users_backend(
     invitee_emails_raw: Annotated[str, ApiParamConfig("invitee_emails")],
     notify_referrer_on_join: Json[bool] = True,
     stream_ids: Json[list[int]],
+    welcome_message_custom_text: Annotated[
```
Sure, posted on CZO: [#api design > customize welcome bot message - #27663 @ 💬](https://chat.zulip.org/#narrow/channel/378-api-design/topic/customize.20welcome.20bot.20message.20-.20.2327663/near/2213648).

saubhagya-patel commented:
@sahil839, I have updated the code following the review received.
To address [this](https://github.com/zulip/zulip/pull/32862#discussion_r2184949059) comment, I have posted on CZO: [#api design > customize welcome bot message - #27663 @ 💬](https://chat.zulip.org/#narrow/channel/378-api-design/topic/customize.20welcome.20bot.20message.20-.20.2327663/near/2213648).
Thank you.

saubhagya-patel commented:
> Does this implement the suggestion by Tim here [#api design > customize welcome bot message - #27663 @ 💬](https://chat.zulip.org/#narrow/channel/378-api-design/topic/customize.20welcome.20bot.20message.20-.20.2327663/near/2213892)?

@sahil839, yes, now this PR implements the mentioned suggestion. Ready for your review.

saubhagya-patel commented on zerver/tests/test_invite.py:
```diff
@@ -2164,6 +2172,132 @@ def test_confirmation_link_in_manual_license_plan(self) -> None:
+        self.assertNotEqual(second_to_last_message.sender.email, "welcome-bot@zulip.com")
+        self.assertEqual(received_initial_direct_message.sender.email, "welcome-bot@zulip.com")
+        self.assertTrue(
+            received_initial_direct_message.content.startswith("Hello, and welcome to Zulip!")
```
Added the test case.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -18840,6 +18936,16 @@ paths:
+
+                          The default value of Welcome Bot custom message provided by the
+                          Administrators to be included in the initial message sent by the
+                          Welcome Bot to new users upon invitation.
```
> This organization's configured custom message for Welcome Bot to send to new user accounts.

(I think the old docs are wrong, now that it's two messages?)

timabbott commented on zerver/lib/onboarding.py:
````diff
@@ -88,6 +94,13 @@ def send_initial_direct_message(user: UserProfile) -> int:
+            welcome_bot_custom_message_string = _("""
+The administrators for this organization would also like to share the following information:
+```quote\n{}\n```
+""").format(welcome_message_custom_text)
````
I would prefer to have the block-quote syntax be outside the translated string. This will save translators work if we decide to make any tweaks to the formatting later.

timabbott commented on zerver/actions/create_user.py:
```diff
@@ -312,10 +312,15 @@ def process_new_human_user(
+        if prereg_user.welcome_message_custom_text is not None:
+            welcome_message_custom_text = prereg_user.welcome_message_custom_text
+        else:
+            welcome_message_custom_text = user_profile.realm.welcome_message_custom_text
```
Why is this part only running if `prereg_user is not None`?

timabbott replied:
Yes we should have the realm's default message here.

timabbott commented on tools/lib/capitalization.py:
```diff
@@ -138,6 +138,8 @@
     # Used in invite modal
     r"or",
+    r"Welcome Bot",
+    r"none",
```
Can you post a screenshot of where this appears tagged for translation in the UI?

timabbott commented on web/src/invite.ts:
```diff
@@ -44,6 +44,18 @@ let stream_pill_widget: stream_pill.StreamPillWidget;
+    invitee_emails: string;
+    include_realm_default_subscriptions: string;
+    welcome_message_custom_text?: string;
+};
```
The refactor to introduce this type would be nicely done as a prep commit, then the main commit just adds a thing or two to the type.

saubhagya-patel commented on zerver/views/welcome_bot_custom_message.py:
```diff
@@ -0,0 +1,35 @@
+    # the first value is the message ID of the initial message. The second value is the
+    # message ID of the Welcome Bot custom message (if provided).
+    assert len(message_ids) > 1
+    welcome_bot_custom_message_id = message_ids[1]
```
This is updated in the following push.

sahil839 commented on web/src/invite.ts:
```diff
@@ -117,6 +118,29 @@ function get_common_invitation_data(): CommonInvitationData {
+
+        if (
+            (realm_welcome_message_configured && !send_realm_default_custom_message) ||
+            (!realm_welcome_message_configured &&
```
This becomes quite difficult to read. I think you can form a function like `get_welcome_message_custom_text_value` and then having a separate block for each condition would be better -
```js
function get_welcome_message_custom_text_value() {
    if (realm_welcome_message_configured) {
        if (send_realm_default_custom_message) {
            data.welcome_message_custom_text = JSON.stringify(null)
        } else {
            data.welcome_message_custom_text = welcome_message_custom_text;
        }
        return;
    }
.....
}
```

timabbott commented on tools/lib/capitalization.py:
```diff
@@ -138,6 +138,7 @@
     r"your-organization",
     # Used in invite modal
     r"or",
+    r"Welcome Bot",
```
This should go next to "Notification Bot" and maybe now have a shared comment that these are system bot names.

timabbott replied:
``` diff
diff --git a/tools/lib/capitalization.py b/tools/lib/capitalization.py
index c11116afb7..950765b00b 100644
--- a/tools/lib/capitalization.py
+++ b/tools/lib/capitalization.py
@@ -138,7 +138,6 @@ IGNORED_PHRASES = [
     r"your-organization",
     # Used in invite modal
     r"or",
-    r"Welcome Bot",
     # Units
     r"MB",
     r"GB",
@@ -161,6 +160,8 @@ IGNORED_PHRASES = [
     r"to add a new line",
     # Used in showing Notification Bot read receipts message
     "Notification Bot",
+    # Used in strings around welcome bot custom messages
+    r"Welcome Bot",
     # Used in presence_enabled setting label
     r"invisible mode off",
     # Typeahead suggestions for "Pronouns" custom field type.
```

timabbott commented on web/src/settings_org.ts:
```diff
@@ -358,6 +361,73 @@ function disable_create_user_groups_if_on_limited_plan(): void {
+    };
+
+    const view_message_button_html = render_action_button(args);
+    $("#welcome_message_custom_text_buttons_container").append($(view_message_button_html));
```
You don't need to add the `$()` around `view_message_button_html`

timabbott replied:
https://api.jquery.com/append/ looks like it allows just a string too; but it looks like we have this pattern a lot, so probably we should discuss in #frontend cleaning that up.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -15387,6 +15394,27 @@ paths:
+                      organization's default `welcome_message_custom_text` is configured,
+                      it is used; otherwise, no Welcome Bot custom message is sent.
+
+                    **Changes**: New in Zulip 11.0 (feature level ZF-0992cf).
```
This is really confusingly worded. Also doesn't mark it as nullable.
``` diff
                   example: false
                 welcome_message_custom_text:
                   description: |
-                    A string denoting the custom text to be included
-                    in the initial message sent by the Welcome Bot to
-                    new users upon joining the organization.
+                    Custom message text, in Zulip Markdown format, to be sent by the
+                    Welcome Bot to new users that join the organization via this
+                    invitation.
 
-                    For invites created by Admin users:
+                    Maximum length is 8000 characters.
 
-                    - `None`: the organization's default `welcome_message_custom_text` is used.
-                    - Empty string: no Welcome Bot custom message is sent.
-                    - Non-empty string: the provided string is sent as the Welcome Bot custom message.
-
-                    For invites created by Non-admin users:
+                    Only organization administrators can use this feature; for other
+                    users, the value is always `null`.
 
-                    - The value of `welcome_message_custom_text` is always `None`. If the
-                      organization's default `welcome_message_custom_text` is configured,
-                      it is used; otherwise, no Welcome Bot custom message is sent.
+                    - `null`: the organization's default `welcome_message_custom_text` is used.
+                    - Empty string: no Welcome Bot custom message is sent.
+                    - Otherwise, the provided string is the custom message.
 
                     **Changes**: New in Zulip 11.0 (feature level 416).
                   type: string
+                  nullable: true
                   example: "Welcome to Zulip! We're excited to have you on board."
             encoding:
               invite_expires_in_minutes:
```

timabbott commented on web/src/tippyjs.ts:
```diff
@@ -849,6 +849,24 @@ export function initialize(): void {
+                return undefined;
+            }
+            instance.destroy();
+            return false;
```
This seems like a pattern we have 5 copies of in this file. Can you start a thread in #frontend to discuss if we should clean these up?

patepelo commented:
Many many many thanks for the wonderful work. When can we expect to use this after the merge?

alya commented:
This is expected to be in the 11.0 release for self-hosters (coming soon), and will probably make it to Zulip Cloud this week.
